### ISSUE TITLE
Evaluation of Semantic Search Variables for Case Data Retrieval

### ISSUE THREAD

**Issue by legaltextai** [freelawproject/courtlistener#4277]
**Proposal**: Assess critical variables affecting semantic search quality on legal corpus before implementating semantic search on CL. 

**Key Variables**:
- Embedding model
- Chunking strategy
- Indexing
- Prompt extraction/rewriting
- Reranking
- Summarization

**Initial Focus**: Start with embedding step. Identify suitable open-source embedding model.

**Project Scope**:
_Dataset_: ~50K Supreme Court opinions (>200 words each)
_Text chunking_: Adapt to embedding model capacity (512-32K+ tokens)
_Models_: Evaluate 6-8 open-source, 1-2 commercial embeddings
_Indexing_: HNSW, cosine similarity (m=16, ef_construction=64); optional hybrid search
_Reranking_: Baseline (none), optional Cohere/BGE-reranker comparison
_Evaluation_: Subjective expert assessment, RAGAS, LlamaIndex/Langchain frameworks

_Resources_: Initial models are available [here](helpmefindcase.com); expansion forthcoming. 

Questions and comments are very welcome.

**Comment by mlissner** [comment]
Very cool. Thanks for getting this conversation started. One question I have, before we get to far, is to what extent using Elastic will limit these options. I'm just thinking that it's possible Elastic will only support some of this, so we might want to identify that before we explore things that won't work?

**Comment by legaltextai** [comment]
Seems like they are [flexible](https://www.elastic.co/search-labs/tutorials/search-tutorial/vector-search/generate-embeddings). At least they directly mention SentenceTransformer framework which includes lots of models.

**Comment by legaltextai** [comment]
What makes you so loyal to Elastic? What is their advantage vs. e.g. Postgres for vector storage or keyword search?

**Comment by mlissner** [comment]
What I like about Elastic:

 - It should give us good hybrid search, allowing the front end UI we currently have to work with vector search without too much trouble.
 - It separates search out from the database so that they don't impact each other.
 - It scales horizontally
 - We already have it in place
 - It's not a third party platform that I don't trust to exist when the AI hype slows down

Things like that. Basically, it's what we have and use for search, so it seems like the right place to add more functionality?

**Comment by legaltextai** [comment]
I take it you host ES on your own servers and the cost is not an issue. I think it 's ok to use it for vector / hybrid search. Plus, llamaindex and langchain have libraries to integrate with ES as a vector search. 
https://docs.llamaindex.ai/en/stable/examples/vector_stores/Elasticsearch_demo/

**Comment by mlissner** [comment]
We do, yes. We have a k8s cluster for Elastic with too many servers in it. :)

Great news!

**Comment by legaltextai** [comment]
**Update on the project:** 

For the _subjective_, ie to get your own feel which model performs better, feel free to head [here](http://helpmefindcase.com) to find Supreme Court cases you know very well. 

For the _objective_ assessment, using hit rate and mrr metrics, these are the numbers I got from my evaluation. 
<img width="665" alt="Screenshot 2024-08-07 at 10 49 57 AM" src="https://github.com/user-attachments/assets/8eeddf33-c4f7-45b6-bfe3-ddad8b28bd07">

I used [these](https://docs.llamaindex.ai/en/stable/examples/finetuning/embeddings/finetune_embedding_adapter/) llamaindex libraries. The train and eval datasets were prepared based on random 10 Supreme Court cases. 

My personal preference is a small but mighty [sentence-transformers/multi-qa-mpnet-base-dot-v1 ] (https://huggingface.co/sentence-transformers/multi-qa-mpnet-base-dot-v1) model. 

What I found interesting, if you fine tune a model, even on a small dataset like 10 cases, there is a 15-20% improvement in performance (local:test_model_3 is my fine tuned model based on base-dot-v1). 

Questions and comments are welcome. 

Cheers, 

PS. 
**Hit Rate:**
Hit rate calculates the fraction of queries where the correct answer is found within the top-k retrieved documents. In simpler terms, it’s about how often our system gets it right within the top few guesses.

**Mean Reciprocal Rank (MRR):**
For each query, MRR evaluates the system’s accuracy by looking at the rank of the highest-placed relevant document. Specifically, it’s the average of the reciprocals of these ranks across all the queries. So, if the first relevant document is the top result, the reciprocal rank is 1; if it’s second, the reciprocal rank is 1/2, and so on.

**Comment by mlissner** [comment]
I'm not sure I know SCOTUS law well enough, but I just shared this across X, BlueSky and Threads. I'll report back if we get any responses. Hopefully those will be useful, but I suspect the better way to do it is going to be objective measures like the ones you've begun using.

Thanks for this. It's really interesting.

**Comment by legaltextai** [comment]
Great. Thanks. 

I just found out that Unstructured, in my view one of the best parsing library, uses [sentence-transformers/multi-qa-mpnet-base-dot-v1](https://huggingface.co/sentence-transformers/multi-qa-mpnet-base-dot-v1) for their ['similarity' chunking](https://unstructured.io/blog/chunking-for-rag-best-practices). Another +1 to this mighty embedding model. As I mentioned, it can also be further improved by doing a fine-tuning prior to the embedding of the whole corpus. 

I also tested [voyage-legal-2](https://blog.voyageai.com/2024/04/15/domain-specific-embeddings-and-retrieval-legal-edition-voyage-law-2/), a highly specialized embedding model for legal texts. Its hit_rate and mrr are higher than any other models I tested. But it is closed, and not free. I would still go with open source.

**Comment by halfprice06** [comment]
Pinging here but I also responded to the X thread on this. 

I'm a practicing attorney and have a different perspective on how to evaluate the embeddings models. 

I think a good way to subjectively (maybe objectively?) test these models is to use "Questions Presented" from pending (or maybe past) supreme court opinions as query input and then measure whether they find cases cited to parties and or the SCOTUS itself in its briefs. 

I have been using Sonnet 3.5 as a judge to determine what are the "most important" cases cited in Petitioner and Respondent briefs and then comparing that list of cases against those retrieved by the embeddings models. 

To me this is a very real world benchmark for evaluating these models. 

Just a thought! Happy to contribute to this btw, rising tide lifts all boats. I'm working on a RAG app that uses courtlistner db as the starting point so happy to give back!

**Comment by honeykjoule** [comment]
cool project! 

lawyers tend to use words very carefully, so i wonder what the benefit of semantic search has over simple keyword matching.

a while back i made embeddings of the parenthetical summary dataset and retrieved them by similarity to the user's input.

i showed my app to some lawyer colleagues and found that they ended up searching for very specific legal terms.

the top results usually included the exact words they searched for.

similarity is cool because you can search "labrador" and find case law related to "canine," but i doubt the usefulness of this for lawyers who are searching for specific terms of art.

the next step for my app was to add filters for jurisdiction and date, but i stopped working on it because i thought that semantic search wouldn't be that useful for lawyers who are looking for pinpoints in semantic space but not necessarily near neighbors.

the project has still been simmering in the back of my mind, so i am excited to see someone else working on nlp + court listener

**Comment by halfprice06** [comment]
> cool project!
> 
> lawyers tend to use words very carefully, so i wonder what the benefit of semantic search has over simple keyword matching.
> 
> a while back i made embeddings of the parenthetical summary dataset and retrieved them by similarity to the user's input.
> 
> i showed my app to some lawyer colleagues and found that they ended up searching for very specific legal terms.
> 
> the top results usually included the exact words they searched for.
> 
> similarity is cool because you can search "labrador" and find case law related to "canine," but i doubt the usefulness of this for lawyers who are searching for specific terms of art.
> 
> the next step for my app was to add filters for jurisdiction and date, but i stopped working on it because i thought that semantic search wouldn't be that useful for lawyers who are looking for pinpoints in semantic space but not necessarily near neighbors.
> 
> the project has still been simmering in the back of my mind, so i am excited to see someone else working on nlp + court listener

So, I can tell you that semantic similarity can be quite useful for attorneys, and it's one of the most requested features I've gotten with regard to the web app I'm building, but the attorney user has to learn how to use it for it be useful which is why I don't think semantic search can ever be the default search mode. 

Casetext built this first for attorneys. They call it parallel search. Check it out: 

https://parallelsearch.casetext.com/

From what I can tell, they chunked by sentences and made embeddings. The app instructs the user to search for a sentence that you don't know for sure exists in a legal opinion somewhere, but that you wish it did. 

So user types in a conclusion they wish existed in a legal opinion and the app does semantic similarity to find similar sentences. This works well if there is a case that actually says that somewhere. Which is why when it works it feels like magic, and attorneys are constantly searching that high! 

A couple of attorneys in my firm use this all the time because they have taken the time to understand how it works. Of course the results are quite poor if you just type in a few keywords the way attorneys normally search for caselaw. 

I think the outside of LLM based agentic search where you have multiple loops back to the well to run more searches while an LLM does some kind of analysis in between, some kind of fine tuned embeddings model with a hybrid search algo which combines cosine similarity and BM25 and with a re-ranking step is going to be SOTA for single "shot" search. 

BM25 is just too good in my experience. 

For my RAG app we do embeddings retrieval and BM25 and then evaluate the results with separate LLM calls, then later combine all of the results into one final answer to the user. Expensive but I've found this maximally increases the footprint of the search to unearth relevant results.

**Comment by halfprice06** [comment]
For what its worth, I also think casetext nailed the UI for choosing between semantic and keyword searching in their UI and CL may want to take inspiration. 

I'm certainly copying parts of it. 

https://casetext.com/cases

This is their default search interface that lets you select between "semantic" and keyword. Clicking inside the search box gives a drop down to let the user select and gives some instructions on how search works.

**Comment by legaltextai** [comment]
Thank you for comments @halfprice06 and @honeykjoule . It 's great to have a group of like minded individuals working on similar projects. 

@honeykjoule  I could prepare another train and eval set based on your approach. If I understand correctly, go through say random 100 SCOTUS cases, extract 'questions' (from questions presented or equivalent) and then 'cited cases' or 'truth' (the cases where the SCOTUS found answers to those questions). Correct? That's a good approach, as we have the ultimate judge (SCOTUS) for retrieval.  

Semantic search alone can be useful exactly for the reason you described. "A dog bit a child" may retrieve "a cat attacked a 3-year old boy". Yes, there are cases about vicious cats :-) Depending on the facts of course, the latter could be useful to analyze the scope of dog owner's liability too. 

@halfprice06 I normally chunk based on the limits of the embedding model. So, if UAE or Cohere have 512 token limit, I chunk by 300-350 word and then embed each chunk. You can see how it works in practice [here](https://helpmefindcase.com/._find_case_by_facts) (UAE/Large embedding model, hnsw index, cosine similarity, with Cohere reranking). I do agree that adding a keyword search should improve the retrieval: 1) you can implement a hybrid search and play with weights b/w semantic and keyword to see which one works best for your needs, and 2) you can give users the option to choose semantic or keyword only. 

@honeykjoule  Take a look at how I implemented the jurisdiction filter. Let me know if you are interested and I can share the code. 

Just to remind us, this is the first step towards the best retrieval for legal texts. Finding a 'good enough' embedding model first, to use as it is or with some fine - tuning. Then we can work on other variables. See my first message above. 

If there is an interest, we can discuss and agree on the best embedding model, fine tuning, and then release all embeddings as open source. As @halfprice06 said, 'rising tide lifts all boats.'

**Comment by halfprice06** [comment]
@legaltextai that was me that suggested the idea of looking at SCOTUS question presented for queries, and yes that's the general idea. 

Since I made the comment I have been thinking about it a little further - I think the dataset needs a good bit of "curation" to be really good, because not every case cited in any particular SCOTUS opinion is actually directly related to the question presented. SCOTUS especially will talk about all kinds of things not directly relevant to the question presented before they finally get to the meat and potatoes. 

But some of the cases (a lot?) that they discuss _will_ be directly relevant. So choosing which cases from the SCOTUS opinions to use as 'cited truth' to grade the retrieval of the embeddings models would be important. In other words, I don't think you can blindly choose which cases are cited in a SCOTUS opinion as your "correct" results, because some of the chunked references to those cited opinons will absolutely not be semantically similar. Somehow you'll need to determine if the cited being directly relevant _semantically_ to the question presented. 

Does that make sense? 

Last thing, something to think about, sometimes SCOTUS cases will for the most part reference the Circuit Courts and not previous SCOTUS opinions, for example when they are trying to resolve Circuit split. So if you don't have the circuit courts embedded yet as part of the retrieval system, it won't be possible to measure the accuracy of the model if you are using Circuit Court cases as "ground truth" cases. 

If it made sense I'd be open to helping with the curation of such a dataset, as I've been planning to make something like this for a while anyway but for the Louisiana courts.

**Comment by legaltextai** [comment]
I think the "SCOTUS approach" might be useful to set the "golden standard" for retrieval.  

For embedding model evaluation, just a step back, what [llamaindex](https://docs.llamaindex.ai/en/stable/examples/finetuning/embeddings/finetune_embedding_adapter/) train / eval library does is it generates questions about each chunk or the whole document (court opinion in our case). It connects the question and the relevant chunk in the train and eval set. Then when you run evaluation of the embedding model it compares whether those pairs, post embedding, match the evaluation set. That's how you get those objective metrics like hit_rate and mrr.  

IMHO, as far as embedding model is concerned I would choose a 'good enough' model with possible fine tuning.  And then focus on the best retrieval model by playing with hybrid search, query decomposition, reranking, etc. to achieve whatever we decide the golden standard should be.  

On a slightly different subject, if anyone is interested in building a framework for shepardizing cases, I 'd be happy to share ideas and collaborate.

**Comment by halfprice06** [comment]
'''

For embedding model evaluation, just a step back, what llamaindex train / eval library does is it generates questions about each chunk or the whole document (court opinion in our case). It connects the question and the relevant chunk in the train and eval set. Then when you run evaluation of the embedding model it compares whether those pairs, post embedding, match the evaluation set. That's how you get those objective metrics like hit_rate and mrr.

'''

Ok that's pretty neat I was wondering where your data set for those metrics came from. 

As good as some of the llm models are now, I do wonder though whether they are really capable of coming up with good questions that are similar to real user queries?  

And I guess that's also a design question too, because evaluating this way is tuning the system specifically for question answering, which is not the only way to do semantic retrieval. 

Questions are often semantically dissimilar from their "correct" answer and so it's interesting way to fine tune.  

With casetext's parallel search, I assume they did not fine tune it for question answering, because the user is not encouraged to input a question. 

My app is designed for users to ask questions and I'm constantly surprised that sometimes users are typing in short phrases or keywords instead of forming questions or even full sentences. 

But all that being said I think that's a good approach to get to the "good enough" stage.

Btw I don't know what would be involved in fine tuning UAE-Large-V1 but in my non scientific  testing it has seemed to do the best in my app for question answering type retrieval of the handful of open models I tried.

Regarding Voyager-legal-2, I actually helped contribute to their data set curation, I recommended a bunch of data and sites for them to scrape in emails with one of the founders. Not surprised to hear it did so well in your tests.  They gave me some platform credits for my help but I never used them because I didn't want to get locked into a closed model.

Also, what about Colbert?

https://blog.vespa.ai/announcing-long-context-colbert-in-vespa/

https://huggingface.co/colbert-ir/colbertv2.0

Finally, regarding building a case citator/shepardizer, I'm down! Something I was planning to do at some point out of necessity anyway.

**Comment by legaltextai** [comment]
```With casetext's parallel search, I assume they did not fine tune it for question answering, because the user is not encouraged to input a question.

My app is designed for users to ask questions and I'm constantly surprised that sometimes users are typing in short phrases or keywords instead of forming questions or even full sentences.```

We can always add a prompt decomposition step to turn a short phrase, keywords, etc into 2-3 relevant questions -> get semantically similar excerpts for each of those -> rerank. 

UAE/Large is good, and I've been using it in my app too, but through this exercise I came across [sentence-transformers/multi-qa-mpnet-base-dot-v1](https://huggingface.co/sentence-transformers/multi-qa-mpnet-base-dot-v1) and like it, especially if you fine-tune it on 100-200 court cases. 

I have not tried Colbert yet. Looks interesting. 

I 've spent $40 on Voyage :-( 

On the case citator / shepardizer, I see two options: 

1) when user clicks to shepardize, we crawl through relevant links in the Courtlistener Citation Map file -> one, two more steps to extract the opinions from CL -> fast LLM reads and summarizes how the case in question was treated. It won't be fast and may be costly depending on the model, but still an option.  

2) extract from CL say 010 , 020 opinions, with > 200 words (those are most likely to discuss other cases) -> get the model to go through each opinion and extract whatever information we need, incl cases cited and how treated. Get the output in a table or json. The good thing about option #2 is that we can use this as an opportunity to turn all those opinions into a structured format, with fields like facts, rule, reasoning, cited statutes and cases, etc. I am sure there will be great opportunities to use that data later for model training, some form of regression and academic  research.  7-8 mln cases. May take 2-3 weeks. I did it last year to extract facts only.

**Comment by s-taube** [comment]
@legaltextai do we need to keep this open, or as this been captured in other issues?

**Comment by legaltextai** [comment]
can we keep it still open under semantic search project?

**Comment by mlissner** [comment]
We can, yes, but what are we using this ticket for at this point that's not broken into other issues?

**Comment by legaltextai** [comment]
i was thinking more of a reference knowledge, a place to add more test results from embedding models

**Comment by legaltextai** [comment]
We did some additional evaluation of embedding models. Previous evals of models with up to 512 token context length are presented above. 

Here are the findings and recommendations from this round of testing: 

_Questions and criteria to consider when choosing embed model:_ 

- Relatively high on [MTEB leaderboard](https://huggingface.co/spaces/mteb/leaderboard?task=retrieval), or retrieval/law category 
- We are going to make embeddings publicly available. Because users will run their own queries, they will need to use the same embed model to embed queries. So, it needs to be affordable. The model will need to run under 16gb of GPU memory if it’s open source, or use widely available third party API 
- What’s the max context length we want to use? How does size affect performance? Is bigger better? Is longer context = bigger memory needed? 

_Does larger context improve retrieval results?_ 

1. Some research has been [done](https://www.llamaindex.ai/blog/evaluating-the-ideal-chunk-size-for-a-rag-system-using-llamaindex-6207e5d3fec5) already. Evaluated OpenAI, with a range from 128 to 2048 token context. 1024 was found to be a sweet spot 

2. What about larger chunks? Does the retrieval improve with increasing the chunk size? 

The results of my tests are below. The dataset is random 30 SCOTUS cases. * [Here](https://towardsdatascience.com/ranking-evaluation-metrics-for-recommender-systems-263d0a66ef54) is the description of the eval metrics.

OpenAI [embed model small](https://platform.openai.com/docs/guides/embeddings/what-are-embeddings), 2000 token chunk size. 
<img width="700" alt="Screenshot 2024-09-26 at 12 54 39 PM" src="https://github.com/user-attachments/assets/a56587bf-2735-4d61-84c8-906059165a7e"> 

Ditto, 4000 token chunk size  
<img width="751" alt="Screenshot 2024-09-26 at 12 55 34 PM" src="https://github.com/user-attachments/assets/4b857619-2544-4156-97b0-527ddd51d647">

Ditto, 8000 token chunk size (the maximum size for this model)  
<img width="756" alt="Screenshot 2024-09-26 at 12 56 05 PM" src="https://github.com/user-attachments/assets/a2252785-10fb-4a64-afb7-7202781c6359"> 

3. Is 8K token context size enough? 

8000 tokens ~ 6000 words ~ 12 pages  

Average number of words in decisions from [harvard cold dataset](https://huggingface.co/datasets/harvard-lil/cold-cases) is 5588. Thus, 8K tokens/6K words should cover most of the decisions. If we don't want to split the texts, and keep one embedding per opinion_id, one approach may be to embed the first 6K words of the decision which should be sufficient to capture the semantic meaning of the facts and issues of the case.  

OpenAI is relatively cheap, widely available, and offers relatively long context. 

_How much better OpenAI’s text-embedding-large vs text-embedding-small on legal data ?_  

Large model is more dimensions (3072 vs 1056) , will take longer to embed , take more space , and is a little bit worse in performance 
<img width="758" alt="Screenshot 2024-09-26 at 1 02 05 PM" src="https://github.com/user-attachments/assets/bfdf054a-7443-4fc0-a4ce-81af1574682a">


I tried two other models with a much larger context togethercomputer/m2-bert-80M-32k-retrieval and m2-bert-80M-8K-retrieval, they performed Ok on general texts (Paul Graham essays) but very poorly on SCOTUS data 
<img width="754" alt="Screenshot 2024-09-26 at 1 02 30 PM" src="https://github.com/user-attachments/assets/69c27ff9-4d5f-426a-a123-29099ac4e537">

How much will it cost to embed with [OpenAI small embed model](https://openai.com/api/pricing/)? 
24 bln x $0.020 / 1M tokens = $480 , or half of that if we run embeddings in batches 

_My recommendation:_ 
1. If we want to increase context length to 8k -> use OAI embed small model. Possibly do a test first , like we did for 50K scotus cases 
2. Otherwise, use the multi-qa-mpnet-base-dot-v1 embed model that we used for the prototype (512 tokens context length). We ‘ll need to discuss how best to store the chunks in Elastic and merge them under the same opinion_id. 

I am pretty sure as new, more capable, models come to the market, we may want to re-embed everything in 6-12 months. For the first version, our search should be ‘good enough’ and we will continue looking for ways to make it better.

**Comment by mlissner** [comment]
This all sounds good, thank you, @legaltextai! One last thing I'm trying to understand is how Elastic handles chunked contexts. In [their document that you shared](https://www.elastic.co/search-labs/blog/semantic-search-simplified-semantic-text), it says:

> Another option is to use chunking to divide long texts into smaller fragments. These smaller chunks are added to each document to provide a better representation of the complete text. You can then use a nested query to search over all the individual fragments and retrieve the documents that contain the best-scoring chunks.

Have we evaluated that as a better option that only embedding the first X tokens? 

> Average number of words in decisions from [harvard cold dataset](https://huggingface.co/datasets/harvard-lil/cold-cases) is 5588. Thus, 8K tokens/6K words should cover most of the decisions.

Not to be too contrarian, but I'm not sure that it's safe to go from the average to the median like this. Many documents are extremely long (hundreds of pages), and many others are very small (dozens of words), so I really don't know if the average is a good representation of what "most" docs are like. I'm not sure how this affects our strategy, but it jumped out at me.

**Comment by legaltextai** [comment]
I was under the impression that we decided not to go with splitting into chunks. I might have misunderstood. 
Are you ok to use [this](https://www.elastic.co/search-labs/blog/semantic-search-simplified-semantic-text 
) ES native embed task approach?

**Comment by mlissner** [comment]
We did decide that, but it was before I saw the way Elastic suggests using nested queries, so now I'm wondering if we can decide on the optimal chunk size based on relevance **and** embed the entire documents.

**Comment by legaltextai** [comment]
OK, I'll build an ES index based on our existing search_opinion mapping, just for SCOTUS cases and experiment with this ES library. Openai-small with 8k context may still be our best option based on the metrics above.

**Comment by mlissner** [comment]
Sounds great. Alberto might have some ideas about nested queries too. I think he's played with them!

**Comment by legaltextai** [comment]
so, i 've played with the semantic_text 

these are the resources i used
https://www.elastic.co/guide/en/elasticsearch/reference/master/infer-service-openai.html
https://www.elastic.co/guide/en/elasticsearch/reference/current/semantic-text.html 

first, you create an inference point (in our case with openai embedding model), which creates a task that sends the text to openai embedding api and inserts the vectors back into 'embeddings' field 

we then update the mapping of our existing index with a new field (e.g. "semantic_text_field"): 

```
"semantic_text_field": {
          "type": "semantic_text",
          "inference_id": "openai-embeddings",
          "model_settings": {
            "task_type": "text_embedding",
            "dimensions": 1536,
            "similarity": "cosine",
            "element_type": "float"
          }
        },
```
    
we then copy the content of 'text' (which is court opinions) from our index into this field which triggers the embedding process

if the text is longer than 250 words, it breaks it into chunks, with embeddings created for each chunk. 

the search is then run with 

```
GET opinion_index/_search
{
  "query": {
    "semantic": {
      "field": "semantic_text_field",
      "query": "cases dealing with second amendment"
    }
  }
}
``` 

note, you don't need to embed the query, it's done automatically by sending it to 'semantic_text_field' 

There are these limitations on the use of semantic_text

Limitations
[edit](https://github.com/elastic/elasticsearch/edit/8.15/docs/reference/mapping/types/semantic-text.asciidoc)
semantic_text field types have the following limitations:
semantic_text fields are not currently supported as elements of [nested fields](https://www.elastic.co/guide/en/elasticsearch/reference/current/nested.html).
semantic_text fields can’t be defined as [multi-fields](https://www.elastic.co/guide/en/elasticsearch/reference/current/multi-fields.html) of another field, nor can they contain other fields as multi-fields.
[6:40](https://flp-talk.slack.com/archives/D07HX6NG7C2/p1727736003394349)
https://www.elastic.co/guide/en/elasticsearch/reference/current/semantic-text.html

this semantic_text protocol takes a bit of time to fine tune to work with our existing index. it would also mean we insert he embeddings into elastic first, and then copy into s3. another option is just to embed everything outside elastic, put into s3 and then copy into our existing index. 

@albertisfu what's your view , given the above limitations, if we can add this new field and run the embedding task in opinion_index? or do you think we should create a separate index for semantic search only?

**Comment by albertisfu** [comment]
> @albertisfu what's your view , given the above limitations, if we can add this new field and run the embedding task in opinion_index? or do you think we should create a separate index for semantic search only?

Regarding the limitations of `semantic_text` with `nested_fields` or `multi-fields`, I don't see an issue since I understand we're only embedding the `text` field in the `opinion_index`, correct? This is just a simple `TextField`.

This brings up another question: Should we include other fields in the `infer_field` embedding, like `case_name`, or is only the opinion text relevant for this semantic search use case?


As for whether we should use the same `opinion_index` or a different index, my only concern with using `opinion_index` is how embedding generation during the indexing process might affect the overall indexing time. Do you know if the embedding process takes place within the same ingestion pipeline, or can it be done in a separate step? I'm worried the embedding process could slow down indexing, especially when ingesting large batches of documents, such as during scrapes. This issue would become more significant when we start running Opinion Search Alerts using the percolator, where fast indexing will be important for sending real-time alerts.

If the impact on indexing performance isn't too severe, I believe having just one index would be ideal. However, if generating embeddings at indexing time is costly, it might be better to maintain a separate index.

In this [tutorial](https://www.elastic.co/guide/en/elasticsearch/reference/master/semantic-search-semantic-text.html), I saw that they use the `re_index` API to generate embeddings from the source index into the target index, which also triggers the embedding process. If generating embeddings during indexing is too resource-intensive, this could be an option, allowing us to generate embeddings once a day based on documents that were indexed or updated during the day. However this will require a especial index only for semantic search.

Regarding Mike's question:
> Sounds great. Alberto might have some ideas about nested queries too. I think he's played with them!

My understanding of `semantic_text` usage is that Elasticsearch handles the generation of chunks for both indexing and searching under the hood. There's no need to use custom nested queries to search within the semantic field, which may have been a solution in the past but is now addressed by `semantic_text`. Right @legaltextai?

**Comment by legaltextai** [comment]
@albertisfu the best would be to give you an access to my replica of opinion_index with embeddings and let you play with it to make sure the presence of embeddings in the same index won't affect the main search. i will follow up via slack

**Comment by mlissner** [comment]
The more I think about it, the more it seems like doing this in a separate process is best: 

 - We use several fields to hold the text of the doc. Elastic won't be smart enough to choose the correct one.
 - Automatically generating embeddings is something that costs money. Seems like the kind of thing that could go badly if, for example, we're updating a record in the DB.
 - Using Elastic for this means we can't deal with performance issues as cleanly since Elastic is doing things aside from just responding to queries.
 - This means the embeddings aren't in S3 or even our DB, but are in Elastic, which isn't the design we want.

**Comment by albertisfu** [comment]
@legaltextai I did some testing on your replica.

I checked the following indices:

```
opinion_index
opinion_index_embed
opinion_index_semantic
```

It seems that in `opinion_index`, you don't have embeddings generated, correct?
In `opinion_index_embed`, it looks like you only have embeddings, with no other fields indexed.
In `opinion_index_semantic`, you have both the regular fields and the generated embeddings. I tested queries in this index, but they're not working properly due to the following issue:

It appears that the mapping for this index was autogenerated by elasticsearch when the documents were copied. As a result, the mapping didn't retain the necessary settings required for regular search, as in the production environment.

- The `cluster_child` field was not properly defined as a `join` field.
- The `raw` field versions were not generated.
- Text fields that support highlighting didn't get the correct setting for vector highlighting: `"term_vector": "with_positions_offsets"`.

To fix this, the index mapping should be explicitly defined at the time of index creation, before copying any documents.

[opinion_index_mapping.txt](https://github.com/user-attachments/files/17234795/opinion_index_mapping.txt)

Once this issue is resolved, the `semantic_text` field can be added.

Aside from this issue with the regular search, I agree with Mike that it might be better to use a different approach than `semantic_text`, especially since embeddings will be generated automatically, as mentioned in docs:

> This means there's no need to create an inference pipeline to generate the embeddings. Using bulk, index, or update APIs will do that for you automatically.

We perform a lot of updates on documents, and if embedding re-computation is triggered on every update for both parent and child documents, it could be quite costly. It would be better to control embedding generation outside of Elasticsearch, on a more reasonable schedule that balances data up-to-date and cost.

If we don't use `semantic_text`, could we still use `sparse_vector` or `dense_vector` fields to generate vectors outside of elasticsearch, and then update the documents with their embeddings using one of those fields?

**Comment by legaltextai** [comment]
Thanks Alberto. What do you get when you run: 

GET opinion_index/_search
{
  "query": {
    "semantic": {
      "field": "semantic_text_field",
      "query": "cases dealing with second amendment"
    }
  }
}

**Comment by albertisfu** [comment]
Yes, I got results from that semantic query. So only issue now is related to the mapping required to support the regular queries we have in production.

**Comment by legaltextai** [comment]
so, just make sure that we update the mapping correctly if we were to add semantic_text to our existing opinion_index? which sounds like we are not going to do anyway
